kge/util/distribution-shift/custom_distribution.py

```diff
@@ -0,0 +1,88 @@
+import torch
+import random
+import os
+import distribution_shift_utils as dsutils
+
+
+class CustomDistribution:
+    def __init__(self, checkpoint_path, noise_magnitude):
+        _, self.dataset = dsutils.load_model_and_dataset(checkpoint_path)
+        self.noise_magnitude = noise_magnitude
+
+    def calculate_target_distribution(self, triples):
+        """Calculate a slightly shifted target distribution by adding random noise."""
+
+        relation_triples = dsutils.group_triples(triples, group_type="relation")
+        original_distribution = {relation: 1.0 for relation, _ in relation_triples.items()}
+        
+        # Add small random noise to the original distribution
+        target_distribution = {
+            relation: max(0, prob + random.uniform(-self.noise_magnitude, self.noise_magnitude))
+            for relation, prob in original_distribution.items()
+        }
+        
+        # Normalize the distribution to ensure it sums to 1
+        total = sum(target_distribution.values())
+        return {relation: prob / total for relation, prob in target_distribution.items()}
+
+    def sample_training_validation_set(self):
+        """Samples the training and validation set based on the target distribution."""
+
+        all_triples = torch.cat((self.dataset.split("train"), self.dataset.split("valid")),dim=0)
+        relation_triples = dsutils.group_triples(all_triples, group_type="relation")
+        target_distribution = self.calculate_target_distribution(all_triples)
+
+        # Calculate target number of triples per relation
+        target_counts = {relation: int(len(all_triples) * prob) for relation, prob in target_distribution.items()}
+
+        # Sample triples for each relation
+        train_valid_triples = []
+        for relation, count in target_counts.items():
+            available_triples = relation_triples[relation]
```

Comment:
Ensuring Triple Availability for Relations:
In line 26 of custom_distribution.py, you ensure that the number of sampled triples doesn’t exceed the available triples. However, it’s also crucial to leave at least one triple per relation in the training set. This is necessary to create embeddings for each relation. Without this, we’d effectively switch to a different KGE setting (inductive link prediction), where there are unseen entities or relations in the test set.